### PR TITLE
add partial rao-blackwellization to trace_elbo

### PR DESCRIPTION
**Pull request by martinjankowiak**
a coarser level of rao-blackwellization than the one in tracegraph_elbo, namely the one that obtains from only taking advantage of some iarange info

### CODE REVIEW

File: tests/infer/test_gradient.py [hunk 1]
```diff
@@ -6,27 +6,74 @@
 import pytest
 import torch
 import torch.optim
-from torch.autograd import Variable
+from torch.autograd import variable
```
Comment (fehiepsi) [inline tests/infer/test_gradient.py]: hi @martinjankowiak , after #860, it is better to use `torch.tensor` everywhere ;)

File: pyro/infer/trace_elbo.py [hunk 2]
```diff
@@ -9,6 +8,23 @@
...
+def compute_site_log_r(model_trace, guide_trace, target_site):
+    log_r = MultiFrameTensor()
+    stacks = model_trace.graph["iarange_info"]['iarange_stacks']
+    for name, model_site in model_trace.nodes.items():
+        if model_site["type"] == "sample":
+            log_r_term = model_site["batch_log_pdf"]
+            if not model_site["is_observed"]:
+                guide_site = guide_trace.nodes[name]
+                guide_log_pdf, _, _ = guide_site["score_parts"]
+                log_r_term -= guide_log_pdf
```
Comment (fritzo) [inline pyro/infer/trace_elbo.py]: avoid in-place operation; instead
```py
log_r_term = log_r_term - guide_site["batch_log_pdf"]
```
(also it's a tad simpler to use `site["batch_log_pdf"]` than `site["score_parts"][0]`)

Comment (martinjankowiak) [reply pyro/infer/trace_elbo.py]: done

File: pyro/infer/trace_elbo.py [hunk 3]
```diff
@@ -9,6 +8,21 @@
 from pyro.infer.elbo import ELBO
 from pyro.poutine.util import prune_subsample_sites
 from pyro.util import check_model_guide_match, check_site_shape, is_nan
+from pyro.infer.util import MultiFrameTensor
+
+
+# TODO cache some of these computations
+def compute_site_log_r(model_trace, guide_trace, target_site):
```
Comment (fritzo) [inline pyro/infer/trace_elbo.py]: With a tiny bit of refactoring you could reduce from quadratic to linear time. First return `log_r` in this function:
```py
def compute_log_r(model_trace, guide_trace):
    ...your existing code...
    return log_r
```
Then lazily compute and cache `log_r` in `Trace_ELBO.loss_and_grads()`:
```py
log_r = None  # compute lazily
for name, model_site in model_trace.nodes.items():
    if model_site["type"] == "sample":
        ...
    else:
        ...
        if not is_identically_zero(score_function_term):
            if log_r is None:
                log_r = compute_log_r(model_trace, guide_trace)
            log_r_site = log_r.sum_t(guide_site["cond_indep_stack"])
            ...
```

**Comment by martinjankowiak** [comment]
let's just merge will address tomorrow

**Comment by fritzo** [comment]
Can we merge #867, then merge this PR, since this PR is based on #867? Or should I merge this and close #867?

File: tests/infer/test_gradient.py [hunk 4]
```diff
@@ -18,12 +18,11 @@
 logger = logging.getLogger(__name__)
 
 
-# TODO make this pass for trace_elbo
 @pytest.mark.parametrize("reparameterized", [True, False], ids=["reparam", "nonreparam"])
 @pytest.mark.parametrize("subsample", [False, True], ids=["full", "subsample"])
 @pytest.mark.parametrize("trace_graph,enum_discrete",
-                         [(True, False), (False, True)],
-                         ids=["TraceGraph", "TraceEnum"])
+                         [(False, False), (True, False), (False, True)],
```
Comment (fritzo) [inline tests/infer/test_gradient.py]: :+1:
